tools/universe/package_builder.py

```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python3
+
+import base64
+import collections
+import difflib
+import hashlib
+import json
+import logging
+import os
+import os.path
+import re
+import tempfile
+
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.DEBUG, format="%(message)s")
+
+_jre_url = 'https://downloads.mesosphere.com/java/jre-8u144-linux-x64.tar.gz'
+_jre_jce_unlimited_url = 'https://downloads.mesosphere.com/java/jre-8u131-linux-x64-jce-unlimited.tar.gz'
+_libmesos_bundle_url = 'https://downloads.mesosphere.io/libmesos-bundle/libmesos-bundle-1.10-1.4-63e0814.tar.gz'
+
+_docs_root = "https://docs.mesosphere.com"
+
+_command_json_filename = 'command.json'
+_config_json_filename = 'config.json'
+_marathon_json_filename = 'marathon.json.mustache'
+_package_json_filename = 'package.json'
+_resource_json_filename = 'resource.json'
+_expected_package_filenames = [
+    _command_json_filename,
+    _config_json_filename,
+    _marathon_json_filename,
+    _package_json_filename,
+    _resource_json_filename]
+
+
+class UniversePackageBuilder(object):
+
+    def __init__(self, package, input_dir_path, upload_dir_url, artifact_paths):
+
+        self._package = package
+        self._upload_dir_url = upload_dir_url
+
+        self.set_input_dir_path(input_dir_path)
+
+        self._artifact_file_paths = {}
+        for artifact_path in artifact_paths:
+            if not os.path.isfile(artifact_path):
+                raise Exception('Provided package path is not a file: {} (full list: {})'.format(
+                    artifact_path, artifact_paths))
+            prior_path = self._artifact_file_paths.get(
+                os.path.basename(artifact_path), '')
+            if prior_path:
+                raise Exception('Duplicate filename between "{}" and "{}". Artifact filenames must be unique.'.format(
+                    prior_path, artifact_path))
+            self._artifact_file_paths[os.path.basename(
+                artifact_path)] = artifact_path
+
+    def set_input_dir_path(self, input_dir_path):
```

Comment:
IIRC, python likes double newlines between functions (here and below). Looks like the previous version didn't have them, but may as well fix it while here...